api/v1alpha1/virtualmachinepublishrequest_types.go

```diff
@@ -0,0 +1,530 @@
+	// request's current state.
+	//
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+func (vmpr VirtualMachinePublishRequest) getCondition(
+	conditionType string,
+) *metav1.Condition {
+
+	for i := range vmpr.Status.Conditions {
+		c := &vmpr.Status.Conditions[i]
+		if c.Type == conditionType {
+			return c
+		}
+	}
+
+	return nil
+}
+
+// IsSourceValid returns true if there is a status condition of Type=SourceValid
+// and Status=True.
+func (vmpr VirtualMachinePublishRequest) IsSourceValid() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionSourceValid)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+// IsTargetValid returns true if there is a status condition of
+// Type=TargetValid and Status=True.
+func (vmpr VirtualMachinePublishRequest) IsTargetValid() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionTargetValid)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+// IsComplete returns true if there is a status condition of Type=Complete
+// and Status=True and the resource's status.completionTime is non-zero.
+//
+// It is recommended to use the MarkComplete function in order to mark
+// a VirtualMachinePublishRequest as complete since the function ensures
+// the status.completionTime field is properly set.
+//
+// This condition is set only when the operation has failed or the
+// VirtualMachineImage resource has been realized.
+func (vmpr VirtualMachinePublishRequest) IsComplete() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionComplete)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return !vmpr.Status.CompletionTime.IsZero()
+	}
+	return false
+}
+
+// IsImageAvailable returns true if there is a status condition of
+// Type=ImageAvailable and Status=True.
+func (vmpr VirtualMachinePublishRequest) IsImageAvailable() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionImageAvailable)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+// IsUploaded returns true if there is a status condition of Type=Uploaded
+// and Status=True.
+//
+// This condition is set when the publication has finished uploading the
+// VM to the image registry and is waiting on the VirtualMachineImage
+// resource to be realized.
+func (vmpr VirtualMachinePublishRequest) IsUploaded() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionUploaded)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+func (vmpr *VirtualMachinePublishRequest) markCondition(
+	conditionType string,
+	status metav1.ConditionStatus,
+	args ...string,
+) {
+	var (
+		message string
+		reason  string
+	)
+	if len(args) > 0 {
+		reason = args[0]
+	}
+	if len(args) > 1 {
+		message = args[1]
+	}
+	if reason == "" && status == metav1.ConditionTrue {
+		reason = "success"
```

Comment:
I might be fussy here but looks like the value of `Reason` should be a CamelCase string, according to https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition